<a href="https://colab.research.google.com/github/FerJGit/TT-2C2025-Data-Analitycs-Notebooks/blob/main/PreEntrega/Mis_Archivos/Pre-entrega.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Análisis de datos con Python - Pre-entrega
Nombre y Apellido: **Fernando Gabriel Jaleh**

# **Etapa 1**

## Script básico para carga de ventas y cálculo de total acumulado
Actividades 2 y 3

In [49]:
# Genero una lista de diccionarios (producto, precio, cantidad) con algunos datos de ejemplo
ventas = [
  {'producto': 'remera', 'precio': 9500, 'cantidad': 3},
  {'producto': 'pantalon', 'precio': 50000, 'cantidad': 1},
  {'producto': 'buzo', 'precio': 30000, 'cantidad': 2}
]

In [50]:
# Calculo total de ventas con map y función lambda
total_ventas = sum(map(lambda venta: venta['precio'] * venta['cantidad'], ventas))

# Imprimo el total con el $ adelante, el punto como separador de miles en lugar de coma, y sin decimales
print(f"El total de ventas es: ${'{:,.0f}'.format(total_ventas).replace(',', '.')}")

El total de ventas es: $138.500


In [48]:
# Calculo total de ventas acumuladas con listas por comprensión
total_ventas = sum([venta["precio"] * venta["cantidad"] for venta in ventas])

# Imprimo el total con el $ adelante, el punto como separador de miles en lugar de coma, y sin decimales
print(f"El total de ventas es: ${'{:,.0f}'.format(total_ventas).replace(',', '.')}")


El total de ventas es: $138.500


## Carga de Datasets
Actividad 1

In [1]:
import pandas as pd

In [2]:
# Importo el dataset ventas desde mi github
df_ventas = pd.read_csv("https://raw.githubusercontent.com/FerJGit/TT-2C2025-Data-Analitycs-Notebooks/refs/heads/main/PreEntrega/Datasets/ventas.csv")

In [3]:
# Importo el dataset clientes desde mi github
df_clientes = pd.read_csv("https://raw.githubusercontent.com/FerJGit/TT-2C2025-Data-Analitycs-Notebooks/refs/heads/main/PreEntrega/Datasets/clientes.csv")

In [4]:
# Importo el dataset marketing desde mi github
df_marketing = pd.read_csv("https://raw.githubusercontent.com/FerJGit/TT-2C2025-Data-Analitycs-Notebooks/refs/heads/main/PreEntrega/Datasets/marketing.csv")

In [42]:
# Veo lo que trae del dataset de ventas
df_ventas

id_venta                producto   precio  cantidad fecha_venta  \
0          792       Cuadro decorativo   $69.94       5.0  02/01/2024   
1          811         Lámpara de mesa  $105.10       5.0  02/01/2024   
2         1156                Secadora   $97.96       3.0  02/01/2024   
3         1372                Heladera  $114.35       8.0  02/01/2024   
4         1546                Secadora  $106.21       4.0  02/01/2024   
...        ...                     ...      ...       ...         ...   
3030      1837         Horno eléctrico  $104.12       9.0  30/12/2024   
3031      2276                  Laptop   $85.27       9.0  30/12/2024   
3032      2696                  Laptop  $107.81       4.0  30/12/2024   
3033      2913              Smartphone   $99.85       7.0  30/12/2024   
3034      2930  Consola de videojuegos   $55.47       6.0  30/12/2024   

              categoria  
0            Decoración  
1            Decoración  
2     Electrodomésticos  
3     Electrodomésticos  
4     Electrodomésticos  
...                 ...  
3030  Electrodomésticos  
3031        Electrónica  
3032        Electrónica  
3033        Electrónica  
3034        Electrónica  

[3035 rows x 6 columns]

In [8]:
# Veo lo que trae del dataset de clientes
df_clientes

id_cliente               nombre  edad         ciudad  ingresos
0             1      Aloysia Screase    44  Mar del Plata  42294.68
1             2  Kristina Scaplehorn    25        Posadas  24735.04
2             3       Filip Castagne    50    Resistencia  35744.85
3             4          Liuka Luard    39   Bahía Blanca  27647.96
4             5        Dore Cockshtt    28        Rosario  28245.65
..          ...                  ...   ...            ...       ...
562         563        Dione Forsyde    29        Posadas  26757.73
563         564          Fleming Gow    39       Santa Fe  43674.96
564         565      Jewelle Mabbett    33        Córdoba  30522.64
565         566          Lauri Munns    23    Resistencia  31259.14
566         567          Micah Matis    31     Corrientes  42927.86

[567 rows x 5 columns]

In [9]:
# Veo lo que trae del dataset de marketing
df_marketing

id_campanha            producto  canal  costo fecha_inicio   fecha_fin
0            74     Adorno de pared     TV   4.81   20/03/2024  03/05/2024
1            12              Tablet   RRSS   3.40   26/03/2024  13/05/2024
2            32     Lámpara de mesa  Email   5.54   28/03/2024  20/04/2024
3            21          Smartphone   RRSS   6.37   29/03/2024  16/05/2024
4            58            Alfombra  Email   4.25   31/03/2024  05/05/2024
..          ...                 ...    ...    ...          ...         ...
85           70          Aspiradora     TV   3.06   13/12/2024  29/12/2024
86           89           Televisor     TV   4.98   13/12/2024    8/2/2025
87           68   Rincón de plantas     TV   5.81   17/12/2024   14/2/2025
88           33            Secadora  Email   3.80   20/12/2024    7/1/2025
89           11  Freidora eléctrica   RRSS   5.27   29/12/2024   21/1/2025

[90 rows x 6 columns]

## Dataframe <font color="red">**ventas**</font>

### Análisis exploratorio de datos
Actividad 4

Aplicamos los métodos de EDA vistos en clase, `info`, `columns`, `dtypes`, `head`, `tail`, `sample`, `unique`, `value_counts`, `describe`, `duplicated`, `isnull`, etc. y documentamos las observaciones:

In [28]:
# Aplicamos info() para familiarizarnos con el dataframe
df_ventas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3035 entries, 0 to 3034
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   id_venta     3035 non-null   int64  
 1   producto     3035 non-null   object 
 2   precio       3033 non-null   object 
 3   cantidad     3033 non-null   float64
 4   fecha_venta  3035 non-null   object 
 5   categoria    3035 non-null   object 
dtypes: float64(1), int64(1), object(4)
memory usage: 142.4+ KB


In [12]:
# Visualizamos los primeros registros con head()
# también podemos usar tail() para visualizar los últimos
df_ventas.head(7)

id_venta           producto   precio  cantidad fecha_venta  \
0       792  Cuadro decorativo   $69.94       5.0  02/01/2024   
1       811    Lámpara de mesa  $105.10       5.0  02/01/2024   
2      1156           Secadora   $97.96       3.0  02/01/2024   
3      1372           Heladera  $114.35       8.0  02/01/2024   
4      1546           Secadora  $106.21       4.0  02/01/2024   
5      1697    Horno eléctrico   $35.35       9.0  02/01/2024   
6      1710   Plancha de vapor   $65.43       2.0  02/01/2024   

           categoria  
0         Decoración  
1         Decoración  
2  Electrodomésticos  
3  Electrodomésticos  
4  Electrodomésticos  
5  Electrodomésticos  
6  Electrodomésticos

In [13]:
# Usar sample() para tomar registros al azar
df_ventas.sample(5)

id_venta    producto   precio  cantidad fecha_venta          categoria
2372      2433  Smartphone  $117.66       5.0  07/10/2024        Electrónica
2550       539  Candelabro   $85.86      12.0  31/10/2024         Decoración
1771      1745    Lavadora   $92.94       8.0  24/07/2024  Electrodomésticos
428       1409    Batidora   $79.10       4.0  20/02/2024  Electrodomésticos
981       2134      Tablet   $43.20       4.0  18/04/2024        Electrónica

In [14]:
# Analizamos el campo precio
# Si es numericos se puede aplicar describe()
df_ventas["precio"].sample(5)

1875     $51.74
188      $26.42
2467    $111.51
1929     $93.10
14       $67.97
Name: precio, dtype: object

In [15]:
# Analizamos el campo cantidad
# si es numerico se puede aplicar describe()
df_ventas["cantidad"].sample(5)

2512     3.0
951      2.0
2577     9.0
1262     8.0
1390    12.0
Name: cantidad, dtype: float64

In [16]:
# Analizamos el campo fecha_venta
df_ventas["fecha_venta"].sample(5)

495     27/02/2024
786     01/04/2024
1971    19/08/2024
2392    10/10/2024
414     18/02/2024
Name: fecha_venta, dtype: object

In [17]:
# Analizamos el campo categoria
# Probar con value_counts y unique
df_ventas["categoria"].value_counts()

categoria
Decoración           1015
Electrónica          1012
Electrodomésticos    1008
Name: count, dtype: int64

### Calidad de datos
Actividad 5

In [18]:
# Buscamos registros duplicados
df_ventas.duplicated().sum()

np.int64(35)

In [19]:
# Visualizamos los duplicados
# keep: 'first', 'last', False
df_ventas[df_ventas.duplicated(keep=False)].sort_values(by="producto")

id_venta         producto   precio  cantidad fecha_venta    categoria
845       300  Adorno de pared   $67.75       9.0  06/04/2024   Decoración
833       300  Adorno de pared   $67.75       9.0  06/04/2024   Decoración
857       345         Alfombra   $56.78       9.0  07/04/2024   Decoración
832       145         Alfombra   $41.72       8.0  06/04/2024   Decoración
868       345         Alfombra   $56.78       9.0  07/04/2024   Decoración
..        ...              ...      ...       ...         ...          ...
854      2108       SmartWatch   $79.37      10.0  06/04/2024  Electrónica
831      2778       SmartWatch   $70.66       7.0  05/04/2024  Electrónica
825      2778       SmartWatch   $70.66       7.0  05/04/2024  Electrónica
877      2373       Smartphone  $123.60      12.0  07/04/2024  Electrónica
866      2373       Smartphone  $123.60      12.0  07/04/2024  Electrónica

[70 rows x 6 columns]

In [20]:
# Buscamos datos nulos
df_ventas.isnull().sum()

id_venta       0
producto       0
precio         2
cantidad       2
fecha_venta    0
categoria      0
dtype: int64

In [22]:
# Visualizar los nulos
df_ventas[df_ventas.isnull().any(axis=1)]

id_venta               producto precio  cantidad fecha_venta    categoria
136       627  Elementos de cerámica    NaN       NaN  17/01/2024   Decoración
139      2171    Parlantes Bluetooth    NaN       NaN  17/01/2024  Electrónica

In [27]:
# Analizar los precios de los NaN a ver que opciones tenemos
# df_ventas[df_ventas["producto"]=="Elementos de cerámica"]
# df_ventas["producto"].unique()
df_ventas["producto"].value_counts()


producto
Lámpara de mesa           181
Auriculares               147
Microondas                135
Cafetera                  118
Laptop                    104
Secadora                  103
Jarrón decorativo         102
Smartphone                102
Alfombra                  102
SmartWatch                102
Aspiradora                102
Cortinas                  102
Rincón de plantas         102
Freidora eléctrica        101
Espejo decorativo         101
Adorno de pared           101
Lavadora                  100
Consola de videojuegos    100
Cuadro decorativo         100
Horno eléctrico           100
Plancha de vapor          100
Proyector                 100
Elementos de cerámica     100
Batidora                  100
Televisor                 100
Cámara digital            100
Parlantes Bluetooth        99
Tablet                     58
Heladera                   49
Candelabro                 24
Name: count, dtype: int64

### Observaciones




* Cantidad de registros: 3035
* Columnas 6
* 35 duplicados, eliminarlos
* Algunos nulos (2), 1% del faltante, dropna
* Campo precio: encontre `$`, removerlo y luego cambiar el dtype a float64
* Campo fecha: ajustar a dtype datetime

## Dataframe <font color="red">**marketing**</font>

### Analisis exploratorio de datos

Aplicamos los métodos de EDA vistos en clase, `info`, `columns`, `dtypes`, `head`, `tail`, `sample`, `unique`, `value_counts`, `describe`, `duplicated`, `isnull`, etc. y documentamos las observaciones:

In [29]:
# Aplicamos info() para familiarizarnos con el dataframe
df_marketing.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 90 entries, 0 to 89
Data columns (total 6 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   id_campanha   90 non-null     int64  
 1   producto      90 non-null     object 
 2   canal         90 non-null     object 
 3   costo         90 non-null     float64
 4   fecha_inicio  90 non-null     object 
 5   fecha_fin     90 non-null     object 
dtypes: float64(1), int64(1), object(4)
memory usage: 4.3+ KB


In [30]:
# Visualizamos los primeros registros con head()
# también podemos usar tail() para visualizar los últimos
df_marketing.head()

id_campanha         producto  canal  costo fecha_inicio   fecha_fin
0           74  Adorno de pared     TV   4.81   20/03/2024  03/05/2024
1           12           Tablet   RRSS   3.40   26/03/2024  13/05/2024
2           32  Lámpara de mesa  Email   5.54   28/03/2024  20/04/2024
3           21       Smartphone   RRSS   6.37   29/03/2024  16/05/2024
4           58         Alfombra  Email   4.25   31/03/2024  05/05/2024

In [44]:
# Analizo el campo producto
df_marketing["producto"].value_counts()

producto
Adorno de pared           3
Tablet                    3
Lámpara de mesa           3
Smartphone                3
Alfombra                  3
SmartWatch                3
Plancha de vapor          3
Batidora                  3
Parlantes Bluetooth       3
Cuadro decorativo         3
Cortinas                  3
Consola de videojuegos    3
Heladera                  3
Jarrón decorativo         3
Lavadora                  3
Televisor                 3
Proyector                 3
Candelabro                3
Espejo decorativo         3
Auriculares               3
Freidora eléctrica        3
Rincón de plantas         3
Cafetera                  3
Aspiradora                3
Elementos de cerámica     3
Cámara digital            3
Secadora                  3
Horno eléctrico           3
Microondas                3
Laptop                    3
Name: count, dtype: int64

In [45]:
# Analizo el campo canal
df_marketing["canal"].value_counts()

canal
TV       30
RRSS     30
Email    30
Name: count, dtype: int64

In [31]:
# Analizamos el campo costo
df_marketing["costo"].sample(5)

17    3.68
52    7.16
28    5.20
55    4.51
70    4.16
Name: costo, dtype: float64

In [32]:
# Analizamos el campo fecha_inicio
df_marketing["fecha_inicio"].sample(5)

10    27/04/2024
80    16/11/2024
68    08/10/2024
73    31/10/2024
1     26/03/2024
Name: fecha_inicio, dtype: object

In [33]:
# Analizamos el campo fecha_fin
df_marketing["fecha_fin"].sample(5)

8     10/05/2024
66    24/10/2024
18    02/07/2024
72    11/12/2024
5     05/05/2024
Name: fecha_fin, dtype: object

### Calidad de los datos


In [36]:
# Busco registros duplicados
df_marketing.duplicated().sum()

np.int64(0)

In [38]:
# Busco nulos
df_marketing.isnull().sum()

id_campanha     0
producto        0
canal           0
costo           0
fecha_inicio    0
fecha_fin       0
dtype: int64

### Observaciones
* Cantidad de registros: 90
* Columnas 6
* No hay duplicados
* No hay nulos
* Campos producto y canal: cambiar el dtype a string
* Campo fecha_inicio y fecha_fin: cambiar a dtype datetime


## Dataframe <font color="red">**clientes**</font>

### Analisis exploratorio de datos

### Observaciones

# Etapa 2

## Limpieza de datos
Actividad 1 - Limpiar el conjunto de datos eliminando duplicados y caracteres no deseados. Documentar el proceso y los resultados.

### Dataframe ventas

* 35 duplicados, eliminarlos - se eliminaron 35 duplicados
* Algunos nulos (2), 1% del faltante, dropna - se eliminaron 2 registros con 2 NaN
* Campo precio: encontre `$`, removerlo y luego cambiar el dtype a float64 - Ok
* Campo fecha: ajustar a dtype datetime - Ok

In [ ]:
# Eliminamos registros duplicados
# atributos a evaluar, keep, ignore_index, inplace
df_ventas.drop_duplicates(inplace=True, ignore_index=True)

In [ ]:
# Validamos que no haya duplicados
df_ventas.duplicated().sum()

In [ ]:
# Eliminamos registros con datos nulos
df_ventas.dropna(inplace=True)

In [ ]:
# Validamos que no haya nulos
# df_ventas.isnull().sum()
df_ventas.info()

In [ ]:
# Eliminamos el caracter especial $ de precio
# Normalización: usar replace()
# De ser necesario, podemos usar el atributo regex=True
df_ventas["precio"] = df_ventas["precio"].str.strip().str.replace("$", "")

In [ ]:
# Validamos que precio no contenga $
df_ventas["precio"].sample(3)

In [ ]:
# Convertimos precio de str a float
# Podemos usar pd.to_numeric o astype o incluso combinarlos
df_ventas["precio"] = df_ventas["precio"].astype(float)

In [ ]:
# Convertir el dtype de cantidad, de float64 a int64
df_ventas["cantidad"]= df_ventas["cantidad"].astype(int)

In [ ]:
# Validar la conversion a int
df_ventas["cantidad"].dtypes

In [ ]:
# Convertir el campo fecha_venta a dtype datetime
# pd.to_datetime(df_ventas["fecha_venta"], format="%d/%m/%Y")
# pd.to_datetime(df_ventas["fecha_venta"], dayfirst=True)
df_ventas['fecha_venta'] = pd.to_datetime(df_ventas['fecha_venta'], format="%d/%m/%Y")

In [ ]:
# Validar la conversion de fecha
df_ventas['fecha_venta'].dtype

### Dataframe marketing

*** Acciones ***

In [ ]:
# Convertir el campo fecha_inicio a dtype datetime


In [ ]:
# Convertir el campo fecha_fin a dtype datetime


In [ ]:
# Validamos los cambios


## Transformación de datos
Actividad 2 - Aplicar filtros y transformaciones para crear una tabla de ventas que muestre solo los productos con alto rendimiento.

In [ ]:
df_ventas.head(2)

In [ ]:
# Transformación:
# Calculamos el valor de cada venta = precio * cantidad
# Agregamos una nueva columnas
df_ventas["valor_venta"] = df_ventas["precio"] * df_ventas["cantidad"]

In [ ]:
# Visualizamos
df_ventas.head(2)

In [ ]:
# Agregación:
# Agrupar valor_venta por producto
# Sumamos las ventas dentro de cada grupo
tabla_ventas = (df_ventas.groupby("producto", as_index=False)["valor_venta"].sum())
tabla_ventas.head()

In [ ]:
# Usemos describe para analizar algunos valores del dataframe resultante
tabla_ventas.describe()

In [ ]:
# Agregamos un plot para visualizar graficamente la distribución
import seaborn as sns
import matplotlib.pyplot as plt

sns.histplot(tabla_ventas["valor_venta"], bins=30, kde=True, edgecolor="black")
plt.title("Distribución de Valor de Venta con Curva de Densidad")
plt.xlabel("Valor de venta")
plt.ylabel("Frecuencia")
plt.show()

In [ ]:
# Calcular el percentil 75 (umbral de alto rendimiento)
umbral_ventas = tabla_ventas["valor_venta"].quantile(0.75)
umbral_ventas

In [ ]:
# Identificar productos que superan el umbral
# productos_top = tabla_ventas.query("valor_venta >= @umbral")
productos_top = tabla_ventas[tabla_ventas["valor_venta"] >= umbral_ventas]
productos_top

## Agregación
Actividad 3 - Resumir las ventas por categoría de producto y analizar los ingresos generados.

In [ ]:
df_ventas.columns

In [ ]:
# Agrupamos por categoria y agregamos la suma
ventas_categoria = df_ventas.groupby("categoria", as_index=False)["valor_venta"].sum()
ventas_categoria.head()

In [ ]:
ventas_categoria

In [ ]:
# Ordenar y mostrar los resultados
ventas_categoria

## Integración de datos
Actividad 4 - Combinar los sets de datos de ventas y marketing para obtener una visión más amplia de las tendencias.

In [ ]:
df_ventas.head(2)

In [ ]:
df_marketing.info()

In [ ]:
# Unir los datasets ventas y marketing por columna 'producto'
ventas_marketing = pd.merge(df_ventas, df_marketing, on="producto", how="left")

In [ ]:
# Verificar coincidencias (muestra)
ventas_marketing.head()

### Analizamos el impacto de la campaña

In [ ]:
# Nos quedamos solo con los registros de ventas durante la campaña de marketing
# Seleccionamos solo los campos que nos interesan: "producto", "valor_venta", "costo"


In [ ]:
# Agrupamos por producto y hacemos una agregación por "valor de venta" y "costo"
ventas_marketing.groupby("categoria")[["valor_venta","costo"]].sum()

In [ ]:
# Generamos una nueva columna con la diferencia entre lo vendido y lo invertido en la campaña
